cdf_fabric_replicator/__main__.py

```diff
@@ -13,30 +13,36 @@
 
 def main() -> None:
     stop_event = CancellationToken()
+    worker_list = []
 
     with EventsReplicator(
         metrics=safe_get(Metrics), stop_event=stop_event
     ) as event_replicator:
         event_worker = threading.Thread(target=event_replicator.run)
         event_worker.start()
-
+        worker_list.append(event_worker)
+        
     with TimeSeriesReplicator(
         metrics=safe_get(Metrics), stop_event=stop_event
     ) as ts_replicator:
         ts_worker = threading.Thread(target=ts_replicator.run)
         ts_worker.start()
-
+        worker_list.append(ts_worker)
+        
     with CdfFabricExtractor(stop_event=stop_event) as extractor:
         extractor_worker = threading.Thread(target=extractor.run)
         extractor_worker.start()
+        worker_list.append(extractor_worker)
 
     with DataModelingReplicator(
         metrics=safe_get(Metrics), stop_event=stop_event
     ) as dm_replicator:
         dm_worker = threading.Thread(target=dm_replicator.run)
         dm_worker.start()
-        dm_worker.join()
+        worker_list.append(dm_worker)
 
+    for worker in worker_list:
+        worker.join()
```

Comment:
Might be worth considering setting up the workers before starting any of the Threads (Sequential Setup). Something like this:

```
    with DataModelingReplicator(
        metrics=safe_get(Metrics), stop_event=stop_event
    ) as dm_replicator:
        worker_list.append(threading.Thread(target=dm_replicator.run))

    for worker in worker_list:
        worker.start()

    for worker in worker_list:
        worker.join()
```

Here is some of the reasoning for doing this...

1. Sequential Setup vs. Dynamic Setup:
- Sequential Setup: You can set up all your thread workers before starting any of them. This approach ensures that all threads are ready to execute their tasks before any actual work begins.
- Dynamic Setup: Alternatively, you can start setting up threads as needed, dynamically creating and launching them during runtime. This approach allows flexibility but may require additional synchronization.
2. Benefits of Setting Up All Threads First:
- Predictable Initialization: When you set up all threads upfront, you ensure that all necessary resources (such as locks, queues, or shared data structures) are properly initialized.
- Avoid Race Conditions: By setting up threads first, you minimize the chances of race conditions during initialization.
- Synchronization: If your threads depend on each other (e.g., one thread produces data, and another consumes it), setting them up sequentially can simplify synchronization.

Reply:
I like the idea of waiting for all threads to complete before finishing

Reply:
I changed to have the Threads started and joined based on your comments but within the same thread which was a good suggestion from @nicholasperkins1992